src/stream/src/executor/source.rs

```diff
@@ -145,23 +145,50 @@ struct SourceReader {
 impl SourceReader {
     #[try_stream(ok = StreamChunkWithState, error = RwError)]
     async fn stream_reader(
-        stream_reader: Arc<Mutex<Box<SourceStreamReaderImpl>>>,
+        stream_reader: Arc<Mutex<Option<Box<SourceStreamReaderImpl>>>>,
         notifier: Arc<Notify>,
         expected_barrier_latency_ms: u64,
+        mut abort_notifier: UnboundedReceiver<()>,
     ) {
         'outer: loop {
             let now = Instant::now();
 
             // We allow data to flow for `expected_barrier_latency_ms` milliseconds.
             while now.elapsed().as_millis() < expected_barrier_latency_ms as u128 {
+                let mut reader = stream_reader.lock().await.take().unwrap();
+                let chunk_result: Option<Result<StreamChunkWithState>>;
+
+                {
+                    let chunk_future = reader.next();
+                    let abort_future = abort_notifier.recv();
+
+                    pin_mut!(chunk_future);
+                    pin_mut!(abort_future);
+
+                    match futures::future::select(chunk_future, abort_future).await {
```

Comment:
Why not use `tokio::select!`?

Reply:
`tokio::select!` does not work well in `try_async` block